Civi/RemoteParticipant/EventSubscriber/MailingListSubscriptionSubscriber.php

```diff
@@ -0,0 +1,180 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+declare(strict_types = 1);
+
+namespace Civi\RemoteParticipant\EventSubscriber;
+
+use Civi\Api4\Group;
+use Civi\RemoteParticipant\Event\ChangingEvent;
+use Civi\RemoteParticipant\Event\GetCreateParticipantFormEvent;
+use Civi\RemoteParticipant\Event\GetParticipantFormEventBase;
+use Civi\RemoteParticipant\Event\GetUpdateParticipantFormEvent;
+use Civi\RemoteParticipant\Event\RegistrationEvent;
+use Civi\RemoteParticipant\Event\UpdateEvent;
+use Civi\RemoteParticipant\Event\ValidateEvent;
+use Civi\RemoteParticipant\MailingList\MailingListSubscriptionManager;
+use Civi\RemoteTools\Api4\Api4Interface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+final class MailingListSubscriptionSubscriber implements EventSubscriberInterface {
+
+  private Api4Interface $api4;
+
+  private MailingListSubscriptionManager $subscriptionManager;
+
+  /**
+   * @inheritDoc
+   */
+  public static function getSubscribedEvents(): array {
+    return [
+      GetCreateParticipantFormEvent::NAME => ['onGetCreateParticipantForm', -1],
+      GetUpdateParticipantFormEvent::NAME => ['onGetUpdateParticipantForm', -1],
+      ValidateEvent::NAME => ['onValidateEvent', -1],
+      RegistrationEvent::NAME => ['onRegistrationEvent', \CRM_Remoteevent_Registration::AFTER_PARTICIPANT_CREATION],
+      UpdateEvent::NAME => ['onUpdateEvent', \CRM_Remoteevent_RegistrationUpdate::AFTER_APPLY_PARTICIPANT_CHANGES],
+    ];
+  }
+
+  public function __construct(Api4Interface $api4, MailingListSubscriptionManager $subscriptionManager) {
+    $this->api4 = $api4;
+    $this->subscriptionManager = $subscriptionManager;
+  }
+
+  /**
+   * @throws \CRM_Core_Exception
+   */
+  public function onGetCreateParticipantForm(GetCreateParticipantFormEvent $event): void {
+    $this->addFields($event);
+  }
+
+  /**
+   * @throws \CRM_Core_Exception
+   */
+  public function onGetUpdateParticipantForm(GetUpdateParticipantFormEvent $event): void {
+    $this->addFields($event);
+  }
+
+  public function onValidateEvent(ValidateEvent $event): void {
+    /**
+     * @phpstan-var array<int, string> $groupIds
+     *   Mapping of group ID as integer to group ID as string.
+     */
+    $groupIds = $event->getSubmission()['mailing_list_group_ids'] ?? [];
+    /** @phpstan-var list<int> $allowedGroupIds */
+    $allowedGroupIds = $event->getEvent()['event_remote_registration.mailing_list_group_ids'] ?? [];
+    if ([] !== array_diff($groupIds, $allowedGroupIds)) {
+      $l10n = $event->getLocalisation();
+      $event->addValidationError('mailing_list_group_ids', $l10n->ts('Invalid value'));
```

Comment:
Is the error message (*Invalid value*) specific enough?

Reply:
Actually this error cannot happen when just the form is used because there are only those IDs selectable that are allowed to be selected. Thus I thought a special message isn't necessary.